pkg/supportbundle/collect.go

```diff
@@ -73,70 +74,74 @@ func runCollectors(collectors []*troubleshootv1beta2.Collect, additionalRedactor
 	collectSpecs = collect.EnsureCollectorInList(collectSpecs, troubleshootv1beta2.Collect{ClusterResources: &troubleshootv1beta2.ClusterResources{}})
 	collectSpecs = collect.EnsureClusterResourcesFirst(collectSpecs)
 
-	var allCollectors []collect.Collector
-
 	allCollectedData := make(map[string][]byte)
 
 	k8sClient, err := kubernetes.NewForConfig(opts.KubernetesRestConfig)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to instantiate Kubernetes client")
 	}
 
+	newDesiredCollectors := make(map[reflect.Type][]collect.Collector)
+
 	for _, desiredCollector := range collectSpecs {
 		if collectorInterface, ok := collect.GetCollector(desiredCollector, bundlePath, opts.Namespace, opts.KubernetesRestConfig, k8sClient, opts.SinceTime); ok {
 			if collector, ok := collectorInterface.(collect.Collector); ok {
 				err := collector.CheckRBAC(context.Background(), collector, desiredCollector, opts.KubernetesRestConfig, opts.Namespace)
 				if err != nil {
 					return nil, errors.Wrap(err, "failed to check RBAC for collectors")
 				}
-
-				if mergeCollector, ok := collectorInterface.(collect.MergeableCollector); ok {
-					allCollectors, err = mergeCollector.Merge(allCollectors)
-					if err != nil {
-						msg := fmt.Sprintf("failed to merge collector: %s: %s", collector.Title(), err)
-						opts.CollectorProgressCallback(opts.ProgressChan, msg)
-					}
-				} else {
-					allCollectors = append(allCollectors, collector)
-				}
+				collectorType := reflect.TypeOf(collector)
+				newDesiredCollectors[collectorType] = append(newDesiredCollectors[collectorType], collector)
 			}
 		}
 	}
 
-	foundForbidden := false
-	for _, c := range allCollectors {
-		for _, e := range c.GetRBACErrors() {
-			foundForbidden = true
-			opts.ProgressChan <- e
+	var foundForbidden bool
+	for collectorType, collectors := range newDesiredCollectors {
+		if mergeCollector, ok := collectors[0].(collect.MergeableCollector); ok {
+			newList, err := mergeCollector.Merge(collectors)
+			if err != nil {
+				msg := fmt.Sprintf("failed to merge collector: %s: %s", mergeCollector.Title(), err)
+				opts.CollectorProgressCallback(opts.ProgressChan, msg)
+			}
+			newDesiredCollectors[collectorType] = newList
```

Comment:
Isn't the aim to have one collector after a merge?

Reply:
That proved to be a much larger refactor of the code consider the Collect and CheckRBAC functions also expect that you pass one collector to it at a time.

Reply:
Cool

I'm happy with this compromise approach. Only thing I would ask is to add some comments letting anyone know that this is a stop gap towards a future simpler implementation, however it would look like.